musikcube 0.99.7 (new formula)

clangen opened the pull request:
<!-- Use [x] to mark item done, or just click the checkboxes with device pointer -->

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

clangen commented:
Currently fighting some CI failures I'm not sure how to address:

```
* Dependency 'curl' is provided by macOS; please replace 'depends_on' with 'uses_from_macos'.
* Dependency 'ncurses' is provided by macOS; please replace 'depends_on' with 'uses_from_macos'.
```

However:
* The version of `ncurses` provided by macOS is known to be buggy, and `musikcube` has display issues using the relatively popular `kitty` terminal emulator: https://github.com/kovidgoyal/kitty/issues/3203
* I prefer to use the version of `curl` provided by homebrew because it's newer, and is closer to versions I support on other platforms. If necessary, we an probably use the system-provided version, but I'm not sure what to do about `ncurses`

SMillerDev commented on Formula/musikcube.rb:
```diff
@@ -0,0 +1,49 @@
+  depends_on "taglib"
+
+  def install
+    @job_count = [Hardware::CPU.cores - 2, 1].max
```
This should be left to the build system

clangen replied:
https://github.com/Homebrew/homebrew-core/pull/118938/commits/785f116167f4b1f242bbba336362f6a5d2c97e97

SMillerDev commented on Formula/musikcube.rb:
```diff
@@ -0,0 +1,49 @@
+
+  def install
+    @job_count = [Hardware::CPU.cores - 2, 1].max
+    system "git",
+              "clone",
+              "--branch",
+              "asio-1-24-0",
+              "https://github.com/chriskohlhoff/asio/",
+              "src/3rdparty/asio/"
```
This should use the brewed version

clangen replied:
https://github.com/Homebrew/homebrew-core/pull/118938/commits/785f116167f4b1f242bbba336362f6a5d2c97e97#diff-ca6572917efe0a188596908555ec5c292dacb9eefc0f2969ca50f6a6b9e31a26R11

SMillerDev commented on Formula/musikcube.rb:
```diff
@@ -0,0 +1,49 @@
+              "https://github.com/chriskohlhoff/asio/",
+              "src/3rdparty/asio/"
+    system "cmake",
+              "-DCMAKE_BUILD_TYPE=Release",
```
This should be used from `std_cmake_args`

clangen replied:
https://github.com/Homebrew/homebrew-core/pull/118938/commits/785f116167f4b1f242bbba336362f6a5d2c97e97#diff-ca6572917efe0a188596908555ec5c292dacb9eefc0f2969ca50f6a6b9e31a26R27

clangen commented:
Addressed PR feedback, but still unable to get past issues related to `musikcube` requiring the `brew` versions of `ncurses` and `curl`.

SMillerDev commented on Formula/musikcube.rb:
```diff
@@ -0,0 +1,40 @@
+  depends_on "taglib"
+
+  def install
+    args = std_cmake_args + %w[
+      -DCMAKE_BUILD_TYPE=Release
+    ]
+    system "cmake", ".", *args
+    system "make"
+    system "cmake", "."
+    system "make", "install"
```
```suggestion
    system "cmake", "-S", ".", "-B", "build", *std_cmake_args
    system "cmake", "--build", "build"
    system "cmake", "--install", "build"
```

clangen replied:
`musikcube` has a somewhat complicated build process due to (1) the number of operating systems and versions it supports, and (2) the dynamic nature of the plugin system. 

If at all possible I'd prefer to keep the precise sequence and contents of the `cmake` commands, as they are used to build every variant across every platform. Is this suggested change non-negotiable?

clangen replied:
https://github.com/Homebrew/homebrew-core/pull/118938/commits/b344ea930578a7cf202ca82762ab85bfd399f396

SMillerDev commented on Formula/musikcube.rb:
```diff
@@ -0,0 +1,40 @@
+  end
+
+  test do
+    system "musikcubed", "--start"
+    system "musikcubed", "--stop"
```
Can we check any output here? Is there some status command we can run?

clangen replied:
Updated to ensure the daemon actually spawned correctly by checking for the existence of a lock file. There's not a lot else we can currently do, as musikcube is by and large an interactive frontend app. https://github.com/Homebrew/homebrew-core/pull/118938/commits/e17c6a56dba3558134c0c25863d3c34afbc46406

clangen replied:
Fixed in subsequent commits.

SMillerDev commented on Formula/musikcube.rb:
```diff
@@ -24,17 +24,19 @@ class Musikcube < Formula
 
   test do
     system "musikcubed", "--start"
+    # if there is no lockfile then the daemon was unable to start properly
+    system "head", "/tmp/musikcubed.lock"
```
I think this works
```suggestion
    # if there is no lockfile then the daemon was unable to start properly
    assert_file_exists "/tmp/musikcubed.lock"
```

clangen replied:
Ah, neato. https://github.com/Homebrew/homebrew-core/pull/118938/commits/2b389cdaea3eb13874f644a3255b122e9d3ec4e2

clangen commented:
> musikcube:
>     * musikcube contains conflicting version recursive dependencies:
>         openssl@3, openssl@1.1
>       View these with `brew deps --tree musikcube`.

Can you provide guidance for resolving this discrepancy? I remember now that the original reason I used `asio` as a submodule instead of a hard dependency was that the version of `openssl` it required via homebrew conflicted with different dependencies (e.g. `curl`, `gnutls`and others)... however, earlier in this PR I was asked to use the `homebrew` vendor'd version instead.

clangen commented:
@SMillerDev looks like the bot is going to auto-close this PR due to lack of activity, but if possible I'd like to continue making progress. Can you please see my previous question?

MikeMcQuaid commented on Formula/musikcube.rb:
```diff
@@ -0,0 +1,42 @@
+class Musikcube < Formula
```
> `musikcube contains conflicting version recursive dependencies:`

You'll need to examine the dependency tree and only depend on things using `openssl@3` or `openssl@1.1`, not both.

clangen replied:
Right, so the problem is `asio` has a hard dependency on `openssl@3`, but **all** other dependencies (e.g. `curl`, `libmicrohttpd`, etc) have a dependency on `openssl@1.1`.

This is why I had originally pulled `asio` in as a git submodule, instead of a `brew` dependency -- but I was asked earlier in this thread to change that. Is there some sort of way for me to request a dependency use a different version of a transitive dependency? Or perhaps create a new version of `asio` that depends on `openssl@1.1`?

MikeMcQuaid commented on Formula/musikcube.rb:
```diff
@@ -0,0 +1,42 @@
+  license "BSD-3-Clause"
+  head "https://github.com/clangen/musikcube.git", branch: "master"
+
+  depends_on "ccache" => :build
```
Is this a soft or hard dependency? Pretty sure this cache won't persist.

clangen replied:
Yeah, you're right, this can be removed. Fixed. https://github.com/Homebrew/homebrew-core/pull/118938/commits/8d065567c7dd5eff57977fdd0f131fea43a54cec

MikeMcQuaid commented on Formula/musikcube.rb:
```diff
@@ -0,0 +1,42 @@
+  depends_on "libopenmpt"
+  depends_on "libvorbis"
+  depends_on "ncurses"
+  depends_on "openssl@1.1"
```
Does this use OpenSSL 1.1 specifically or can it use 3?

clangen replied:
As mentioned above, all transitive dependencies (with the exception of `asio`) depend on 1.1.

MikeMcQuaid replied:
@Homebrew/core How are we handling things like this today? Thanks!

carlocab replied:
We haven't really run into an issue like this because the audit prevents it from happening.

Is `asio` a runtime dependency? It's a header-only library, so I suspect it is build-only. Making this `:build` might fix the audit failure.

clangen replied:
Good point, it can almost definitely be defined as a build-time dependency. Let's give that a go... https://github.com/Homebrew/homebrew-core/pull/118938/commits/4c23788433bee799e03c526fd874ee8c96ee412b

clangen replied:
OK, I can't tell if making it a build-time dependency has fixed anything, because I'm seeing the following failure in CI:

```
Full audit musikcube --online --new-formula output
  musikcube:
    * Dependency 'curl' is provided by macOS; please replace 'depends_on' with 'uses_from_macos'.
    * Dependency 'ncurses' is provided by macOS; please replace 'depends_on' with 'uses_from_macos'.
```

However, as previously mentioned, the version of `ncurses` provided by macOS has some really bad bugs that prevent the app from functioning properly in the popular `kitty` terminal emulator, so I'd very much prefer to use the `ncurses` provided version.

Additionally, I'd like to use the most recent version of `curl` simply because it will always have the most recent security patches.

Is it possible to override this particular test failure so I can see if the `openssl` dependency conflict has been solved by recent commits?

p-linnane replied:
Add your formula to https://github.com/Homebrew/homebrew-core/blob/master/audit_exceptions/provided_by_macos_depends_on_allowlist.json in this PR.

clangen replied:
> I'm fine with using the ncurses formula if using the system one leads to bugs.

Cool, is there something I need to do on my end to move past this failure? 

> Apple ships security patches for system `curl`.

Got it. I've updated `curl` to use `uses_from_macos`

> We can still merge this even if there is a brew audit failure for dependencies.

Nice. I temporarily migrated to `uses_from_macos` to `ncurses` to make sure there are no additional build failures, but have reverted back to `depends_on`. Builds are in flight right now, but I'm expecting them to fail again with the same error.

Is there something additional that needs to be done to ensure builds progress pass this audit error so CI is happy and binaries are generated?

MikeMcQuaid reviewed (commented):
Need to ask for Homebrew/core thoughts.

carlocab commented on Formula/musikcube.rb:
```diff
@@ -33,9 +33,9 @@ def install
-    system "musikcubed", "--start"
+    system "#{bin}/musikcubed", "--start"
     # if there is no lockfile then the daemon was unable to start properly
     assert_file_exists "/tmp/musikcubed.lock"
```
```suggestion
    assert_path_exists "/tmp/musikcubed.lock"
```

MikeMcQuaid commented on Formula/musikcube.rb:
```diff
@@ -24,10 +24,16 @@ class Musikcube < Formula
+        "-DNO_NCURSES_W=true",
+        "-DENABLE_MACOS_SYSTEM_NCURSES=true",
+      ]
+    end
```
```suggestion
    platform_specific_args = if OS.mac?
      [
        "-DNO_NCURSES_W=true",
        "-DENABLE_MACOS_SYSTEM_NCURSES=true",
      ]
    else
      []
    end
```

clangen replied:
This was temporary code added to move past the audit failure to ensure there were no additional build problems. It has been reverted.

MikeMcQuaid reviewed (commented):
Deferring to @carlocab here.

carlocab commented on Formula/musikcube.rb:
```diff
@@ -0,0 +1,40 @@
+  uses_from_macos "curl"
+
+  def install
+    system "cmake", ".", *std_cmake_args
+    system "make"
+    system "make", "install"
```
```suggestion
    system "cmake", "-S", ".", "-B", "build", *std_cmake_args
    system "cmake", "--build", "build"
    system "cmake", "--install", "build"
```

carlocab replied:
Yea, it's been on my todo list for a while 😅

clangen replied:
https://github.com/Homebrew/homebrew-core/pull/118938/commits/b344ea930578a7cf202ca82762ab85bfd399f396

carlocab commented on Formula/musikcube.rb:
```diff
@@ -0,0 +1,40 @@
+  end
+
+  test do
+    system "#{bin}/musikcubed", "--start"
```
```suggestion
    system bin/"musikcubed", "--start"
```

MikeMcQuaid replied:
Same here.

clangen replied:
https://github.com/Homebrew/homebrew-core/pull/118938/commits/b344ea930578a7cf202ca82762ab85bfd399f396

carlocab commented on Formula/musikcube.rb:
```diff
@@ -0,0 +1,40 @@
+    system "sleep", "5"
+    assert_path_exists "/tmp/musikcubed.lock"
+    system "sleep", "5"
+    system "#{bin}/musikcubed", "--stop"
```
```suggestion
    system bin/"musikcubed", "--stop"
```

clangen replied:
https://github.com/Homebrew/homebrew-core/pull/118938/commits/b344ea930578a7cf202ca82762ab85bfd399f396

carlocab reviewed (approved):
This is fine with me; just a few more style nits. Also:

Can you rebase against `origin/master` to get rid of the merge commit (which is why you see the https://github.com/Homebrew/homebrew-core/labels/automerge-skip flag), and then squash all your commits together? Thanks!

clangen commented:
> This is fine with me; just a few more style nits. Also:
> 
> Can you rebase against `origin/master` to get rid of the merge commit (which is why you see the [ automerge-skip](https://github.com/Homebrew/homebrew-core/labels/automerge-skip) `brew pr-automerge` will skip this pull request flag), and then squash all your commits together? Thanks!
> 
> Alternatively, I can do this for you, but you need to check the box that says "allow maintainers to edit your PR" (or similar) on the right side.

Addressed remaining feedback and gave you edit access as requested.

carlocab commented:
I'm still not able to push to your branch, so I opened #126426 instead. Let's continue this there.

carlocab commented:
Merged in https://github.com/Homebrew/homebrew-core/commit/c36308094fa9e382dfdd5c014d3d97188491dfb9.

Thanks for the contribution!
